Add pull request filtering in aggregations

pimterry opened the pull request:
This pull request is an attempt to deal with #1043. This is likely to end up closer to being a proposal than a final answer, but I'm hoping a concrete fix might help move conversation forward a bit.

Some notes:

* This changes how PRs appear on the org leaderboard and pages, and the user PR total on user badges.
* It doesn't change the PR total or list of PRs shown on the user page.
* This is a slightly simpler filtering design than proposed at some points, filtering only on repo pattern for a given user. We could easily extend this and tweak it to whatever else we wanted though, with tiny changes in `aggregation_filter.rb`, I'm checking the overall approach first. I think this is a reasonable & effective solution, which I'd be happy to have it merged if people agree, but I'm not too attached to it.
* This includes the removal of `Organisation.with_user_counts`, `User.users_with_pull_request_counts`, `organisation_tooltip`, and a related helper: they'd potentially need changes en route, but fortunately none of this code is actually being used anywhere any more, so I've just got rid of all of it.
* There's no indexing here for the repo name filtering. That can be done but it's non-trivial (because of the non-constant LIKE query). I think it won't matter much anyway since there'll be very very few of these filters and they're scoped by user, so the patterns won't be being checked very often, and this is only ever used in the background job anyway, so small potential delays aren't a big problem.

I've done very little Rails before, so this might not be idiomatic, sorry! Do shout if there are issues, happy to change things to fit The Rails Way.

andrew commented on app/models/aggregation_filter.rb:
```diff
@@ -0,0 +1,7 @@
+class AggregationFilter < ActiveRecord::Base
+  belongs_to :user
+
+  PULL_REQUEST_FILTER = AggregationFilter.where("pull_requests.user_id = aggregation_filters.user_id")
```
This would probably be more obvious as a class method on `AggregationFilter`, something like:

```ruby
def self.pull_request_filter
  where("pull_requests.user_id = aggregation_filters.user_id")
  .where("pull_requests.repo_name LIKE aggregation_filters.repo_pattern")
  .exists.not
end
```

andrew commented on lib/tasks/pull_requests.rake:
```diff
@@ -15,7 +15,7 @@ desc 'Refresh pull request counts'
 task refresh_pull_request_counts: :environment do
   User.reset_column_information
   User.all.each do |u|
```
We can replace this whole block with: 

```ruby
User.all.find_each(&:update_pull_request_count)
```

andrew commented:
Thanks for this @pimterry, I've already pulled in some of your dead code removal changes, if you rebase against master it should simplify the diff a little.

I've also added some inline comments around making things follow the Rails Way®

What do you see the first filter being added as?

pimterry commented:
I've now rebased, and fixed up your comments.

I've also had a good proper look through for a more thorough set of examples, and I think it does look like filtering by title instead of the repo name would work more easily and fairly for many of those, with fewer false matches. I've made the tiny update to switch this to that. Given that, for patterns I'd suggest:

User | PR title pattern
------  | --------
[IonicaBizau](http://24pullrequests.com/users/IonicaBizau) | Improved Documentation
[IonicaBizau](http://24pullrequests.com/users/IonicaBizau) | % Improved Documentation
[sanemat](http://24pullrequests.com/users/sanemat) | Request by PullRequest::Create
[kytrinyx](http://24pullrequests.com/users/kytrinyx) | Switch to container-based Travis CI infrastructure
[kytrinyx](http://24pullrequests.com/users/kytrinyx) | Normalize YAML in Travis CI config

Each of those is based on a whole batch of automated near identical PRs in extremely quick succession recently. I've had a quick flick through the rest of the top of the users list, I think this probably covers it.

andrew commented on app/models/aggregation_filter.rb:
```diff
@@ -0,0 +1,9 @@
+
+  def self.pull_request_filter
+    where("pull_requests.user_id = aggregation_filters.user_id")
+    .where("pull_requests.title LIKE aggregation_filters.title_pattern")
```
We can use `ILIKE` here for case insensitivity

pimterry replied:
Excellent point. Done!

andrew commented:
Merged, deployed, created filters for those three users and updated the pr counts for them and any orgs they belong to, thanks again @pimterry :tada:

IonicaBizau commented:
:+1:

Charlotteis commented:
Good work!
